### PR TITLE
Roslyn syntax transformation

### PR DESCRIPTION
**Pull request by BillWagner**
Adding the next quick start to the series of analyzer tutorials.

Relies on dotnet/samples#107

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/get-started/syntax-transformation?branch=pr-en-us-5837)

### CODE REVIEW

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+This project uses the <xref:Microsoft.CodeAnalysis.CSharp.SyntaxFactory?displayProperty=nameWithType> class methods to construct a <xref:Microsoft.CodeAnalysis.CSharp.Syntax.NameSyntax?displayProperty=nameWithType> representing the `System.Collections.Generic` namespace.
+
+Add the following using directive to the top of the `Program.cs` file to import the factory methods of the <xref:Microsoft.CodeAnalysis.CSharp.SyntaxFactory> class and the methods of <xref:System.Console> so that you can use them later without qualifying them:
+
+[!code-csharp[import the SyntaxFactory class](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/ConstructionCS/Program.cs#StaticUsings "import the Syntax Factory class and the System.Console class")]
+
+You create **name syntax nodes** to build the tree that represents `using System.Collections.Generic;'. <xref:Microsoft.CodeAnalysis.CSharp.Syntax.NameSyntax> is the base class for four types of names that appear in C#. You compose these four types of names together to create any name that can appear in the C# language:
```
Comment (pkulikov) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: nit: it looks that codefencing of "using System.Collections.Generic;" is not closed properly (with a single quote)

**Comment by BillWagner** [comment]
closing and reopening to start a new build. Dependent PR is merged.

**Comment by BillWagner** [comment]
closing and re-opening to force a build.

**Comment by rpetrusha** [comment]
@BillWagner, there are three broken links. The first (IdentifierName) should end with %2A. For the other two, ReplaceNode is not a member of CompilationUnitSyntax.

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+The next step is to create a tree that represents an entire (small) program and then modify it. Add the following code to the beginning of the `Program` class:
+
+[!code-csharp[create a parse tree](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/ConstructionCS/Program.cs#DeclareSampleCode "Create a tree that represents a small program")]
+
+> [!NOTE]
+> The file uses the `System.Collections` namespace and not the `System.Collections.Generic` namespace.
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: The file, or the example?

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+The next step is to create a tree that represents an entire (small) program and then modify it. Add the following code to the beginning of the `Program` class:
+
+[!code-csharp[create a parse tree](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/ConstructionCS/Program.cs#DeclareSampleCode "Create a tree that represents a small program")]
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: perhaps SampleCode should be camel cased?

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+Run the program again. This time the tree now correctly imports the `System.Collections.Generic` namespace.
+
+### Transform trees using `SyntaxRewriters`
+
+The `With*` and <xref:Microsoft.CodeAnalysis.SyntaxNodeExtensions.ReplaceNode%2A> methods provide convenient means to transform individual branches of a syntax tree. The <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter?displayProperty=nameWithType> class performs multiple transformations on a syntax tree. The <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter?displayProperty=nameWithType> class is a subclass of <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor%601?displayProperty=nameWithType>. The <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter> applies a transformation to a specific type of <xref:Microsoft.CodeAnalysis.SyntaxNode>. You can apply transformations to multiple types of <xref:Microsoft.CodeAnalysis.SyntaxNode> objects wherever they appear in a syntax tree. The second project in this quickstart creates a command-line refactoring that removes explicit types in local variable declarations anywhere where type inference could be used.
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: anywhere where --> anywhere that

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+```C#
+public override SyntaxNode VisitLocalDeclarationStatement(LocalDeclarationStatementSyntax node)
+{
+
+}
+```
+
+> [!NOTE]
+> The <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxRewriter.VisitLocalDeclarationStatement(Microsoft.CodeAnalysis.CSharp.Syntax.LocalDeclarationStatementSyntax)> method returns a <xref:Microsoft.CodeAnalysis.SyntaxNode>, not <xref:Microsoft.CodeAnalysis.CSharp.Syntax.LocalDeclarationStatementSyntax>. In many scenarios one kind of node may be replaced by another kind of node entirely - or even removed. In this example you'll return another <xref:Microsoft.CodeAnalysis.CSharp.Syntax.LocalDeclarationStatementSyntax> node based on the existing one.
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: - not --> not a
- comma after scenarios
- comma after example
- The general note is somewhat confusing. Make it clear that SyntaxNode is the base class of nodes, including LocalDeclarationStatementSyntax.

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+This quickstart handles local variable declarations. You could extend it to other declarations such as `foreach` loops, `for` loops, LINQ expressions, and lambda expressions. Furthermore this rewriter will only transform declarations of the simplest form:
+
+```C#
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: C# --> csharp

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+Type variable = expression;
+```
+
+If you want to explore on your own, consider extending the finished sample for these types of variable declarations:
+
+```C#
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: C# --> csharp

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+Finally, add the following `if` statement to replace the existing type name with the `var` keyword if the type of the initializer expression matches the type specified:
+
+[!code-csharp[ReplaceNode](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/TransformationCS/TypeInferenceRewriter.cs#BindInitializer "Replace the initializer node")]
+
+The conditional is required because the declaration may be casting the initializer expression to a base class or interface. If that's desired, the types on the left and right-hand side of the assignment don't match. Removing the explicit type in these cases would change the semantics of a program. `var` is specified as an identifier rather than a keyword because `var` is a contextual keyword. The leading and trailing trivia (whitespace) are transferred from the old type name to the `var` keyword to maintain vertical whitespace and indentation. It's simpler to use `ReplaceNode` rather than `With*` to transform the <xref:Microsoft.CodeAnalysis.CSharp.Syntax.LocalDeclarationStatementSyntax> because the type name is actually the grandchild of the declaration statement.
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: be casting --> cast

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+You've finished the `TypeInferenceRewriter`, now return to your `Program.cs` file to finish the example. Create a test <xref:Microsoft.CodeAnalysis.Compilation> and obtain the <xref.Microsoft.CodeAnalysis.SemanticModel> from it. Use that <xref:Microsoft.CodeAnalysis.SemanticModel> to try your `TypeInferenceRewriter`. You'll do this step last. In the meantime declare a placeholder variable representing your test Compilation:
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: - "Now return..." is a new sentence.
- the first SemanticModel is a broken link: xref. --> xref:
- Compilation should either be a link or, preferably, lowercase (compilation)

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+[!code-csharp[DeclareCompilation](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/TransformationCS/Program.cs#DeclareTestCompilation "Declare the test compilation")]
+
+After pausing a moment, you should see an error squiggle appear reporting that no `CreateTestCompilation` method exists. Press **Ctrl+Period** to open the light-bulb and then press Enter to invoke the **Generate Method Stub** command. This command will generate a method stub for the `CreateTestCompilation` method in the `Program` class. You'll come back to fill in this method later:
+
+![C# Generate method from usage](./media/syntax-transformation/generate-from-usage.png)
+
+Write the following code to iterate over each <xref:Microsoft.CodeAnalysis.SyntaxTree> in the test <xref:Microsoft.CodeAnalysis.Compilation>. For each one initialize a new `TypeInferenceRewriter` with the <xref:Microsoft.CodeAnalysis.SemanticModel> for that tree:
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: comma after "For each one"

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+[!code-csharp[IterateTrees](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/TransformationCS/Program.cs#IterateTrees "Iterate all the source trees in the test compilation")]
+
+Inside the `foreach` statement you created, add the following code to perform the transformation on each source tree. This code conditionally writes out the new transformed tree if any edits were made. Your rewriter should only modify a tree if it encountered one or more local variable declarations that could be simplified using type inference:
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: encountered --> encounters

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+[!code-csharp[TransformTrees](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/TransformationCS/Program.cs#TransformTrees "Transform and save any trees that are modified by the rewriter")]
+
+You should see squiggles under the `File.WriteAllText` code. Select the light bulb, and add the necessary `using System.IO;` statement.
+
+You're almost done! There's once step left. Creating a test <xref:Microsoft.CodeAnalysis.Compilation>. Since you haven't been using type inference at all during this quickstart, it would have made a perfect test case. Unfortunately, creating a Compilation from a C# project file is beyond the scope of this walkthrough. But fortunately, if you've been following instructions carefully there's hope. Replace the contents of the `CreateTestCompilation` method with the following code. It creates a test compilation that coincidentally matches the project described in this quickstart:
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: left. Creating --> left: creating
comma after carefully

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+[!code-csharp[CreateTestCompilation](../../../../samples/csharp/roslyn-sdk/SyntaxTransformationQuickStart/TransformationCS/Program.cs#CreateTestCompilation "Create a test compilation using the code written for this quickstart.")]
+
+Cross your fingers and run the project. In Visual Studio, choose **Debug** > **Start Debugging**. You should be prompted by Visual Studio that the files in your project have changed. Click "**Yes to All**" to reload the modified files. Examine them to observe your awesomeness. Note how much cleaner the code looks without all those explicit and redundant type specifiers.
+
+Congratulations! You've used the **Compiler APIs** to write your own refactoring that searches all files in a C# project for certain syntactic patterns, analyzes the semantics of source code that matches those patterns, and transforms it. You're now officially Refactoring author!
```
Comment (rpetrusha) [inline docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: Refactoring author --> a refactoring author (?)

**Review by rpetrusha — CHANGES_REQUESTED** [review]
This looks good, @BillWanger. I've left a number of comments and suggestions.

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+Run the program again to see that you've build the tree for the code to add.
+
+### Create a modified tree
+
+You've built a small syntax tree that contains one statement. The APIs to create new nodes are the right choice to create single statements, or other small code blocks. However, to build larger blocks of code, you should use methods that replace nodes, or insert nodes into an existing tree. Remember that syntax trees are immutable. The **Syntax API** doesn't provide any mechanism for modifying an existing syntax tree after construction. Instead, the **Syntax API** provides methods that produce new trees based on changes to existing ones. Concrete classes that derive from <xref:Microsoft.CodeAnalysis.SyntaxNode> defines `With*` methods that specify changes to its child properties. Additionally, the <xref:Microsoft.CodeAnalysis.SyntaxNodeExtensions.ReplaceNode%2A> extension method can be used to replace a descendent node in a subtree. This method also updates the parent to point to the newly created child and repeats this process up the entire tree - a process known as _re-spining_ the tree.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: re: last comment. There are some instance methods, and some extension methods.

File: docs/csharp/roslyn-sdk/get-started/syntax-transformation.md (hunk 3 above)
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-transformation.md]: Fixed in dotnet/samples#110

**Comment by BillWagner** [comment]
Thanks @rpetrusha Those all make this topic much more approachable.

All the feedback has been addressed.

**Review by rpetrusha — APPROVED** [review]
This looks really good, @BillWagner. You can merge when you're ready.

**Pull request merged**